File: aoc2024/02/main.py
from functools import reduce
from copy import deepcopy

FILE = "data.txt"
# FILE = "test.txt"

def load_data():
    with open(FILE, "r") as f:
        lines = []
        for line in f:
            lines.append(list(map(int, line.split(" "))))
    return lines

def check_save(report):
    is_inc = None
    diffs = list(map(lambda p: p[0] - p[1], zip(report[0:len(report)-1], report[1:])))
    return all(map(lambda x: 1 <= abs(x) <= 3, diffs)) and (all(map(lambda x: x > 0, diffs)) or all(map(lambda x: x < 0, diffs)))

def search_problem(diffs):
    parity = diffs[0] > 0
    for i in range(len(diffs)):
        diff = diffs[i]
        if parity and diff < 0: return i
        elif not parity and diff > 0: return i

        if not(1 <= abs(diff) <= 3): return i
    return -1

def get_diffs(report):
    diffs = list(map(lambda p: p[0] - p[1], zip(report[0:len(report)-1], report[1:])))
    return diffs

def check_save2(report):
    is_inc = None
    diffs = get_diffs(report)
    for i in range(len(report)):
        report1 = deepcopy(report)
        del report1[i]
        if check_save(report1): return True
    return False

def sol1():
    lines = load_data()
    safe = 0
    for report in lines:
        if check_save(report): safe += 1
    print("Solution 1:", safe)

def sol2():
    lines = load_data()
    safe = 0
    for report in lines:
        if check_save2(report): safe += 1
    print("Solution 2:", safe)

if __name__ == "__main__":
    sol1()
    sol2()

File: aoc2024/02/main.hs

-- filename = "test.txt"
filename = "data.txt"

parseLines :: String -> [[Int]]
parseLines contents = map (\l -> map read l) $ map words (lines contents)

dropLast :: [a] -> [a]
dropLast [] = []
dropLast [x] = []
dropLast lst = (reverse . tail . reverse) lst 

removeNth :: Int -> [a] -> [a]
removeNth n xs =
    let (before, after) = splitAt n xs
    in before ++ drop 1 after

checkSave :: [Int] -> Bool
checkSave lst = (all (\x -> (1 <= (abs x)) && ((abs x) <= 3)) diffs) && ((all (< 0) diffs) || (all (> 0) diffs))
  where
    diffs = map (\(x, y) -> x - y) $ zip (dropLast lst) (tail lst)

checkSave2 :: Int -> [Int] -> Bool
checkSave2 n lst
  | n >= length lst = False
  | otherwise = if (checkSave $ removeNth n lst) then True else checkSave2 (n+1) lst

sol1 :: [[Int]] -> Int
sol1 lists = length $ filter checkSave lists 

sol2 :: [[Int]] -> Int
sol2 lists = length $ filter (checkSave2 0) lists
  
main :: IO ()
main = do 
  contents <- readFile filename
  let lists = parseLines contents
  putStrLn $ "Solution 1: " ++ (show $ sol1 lists)
  putStrLn $ "Solution 2: " ++ (show $ sol2 lists)
  return ()
